### DISCUSSION TITLE
Calculating the DTLS 1.3 transcript hash

### DISCUSSION THREAD

**Discussion by fwh-dc** [openssl/openssl#25803] [General Discussion]
I am studying the way that the transcript hash is calculated for DTLS 1.3.

The problem I am facing is that as per rfc9147: 

> In DTLS 1.3, the message transcript is computed over the
> original TLS 1.3-style Handshake messages without the
> message_seq, fragment_oﬀset, and fragment_length values. Note
> that this is a change from DTLS 1.2 where those values were
> included in the transcript.

Since the client does not know whether DTLS 1.3 will be used when it sends its initial ClientHello I assume that it should keep track of two digests until the version has been negotiated.

But looking at the code I stumbled upon `s3.handshake_buffer`. Does someone care to explain its role?
I was thinking that I could leverage on it instead of maintaining multiple digests.

**Comment by mattcaswell** [comment]
Initially we don't digest the handshake records at all. We just keep a record of all the messages exchanged so far in a mem BIO (`s3.handshake_buffer`). As we send and receive messages we feed them in to the mem BIO via calls to `ssl3_finish_mac`. 

At a certain point in the handshake (and there are various of them) we need to start actually digesting, and so we call `ssl3_digest_cached_records`. Search for calls to that function for all the places where this occurs. Basically that function constructs an `EVP_MD_CTX` (`s3.handshake_dgst`) and feeds all data that we have stored up in the `s3.handshake_buffer` through it.

After that point `ssl3_finish_mac` will stop using the mem BIO, and will feed all messages directly through `s3.handshake_dgst`.

**Reply by fwh-dc** [reply]
Thanks Matt. That helps. I'm wondering if it would be acceptable to keep two record of messages one for "DTLS 1.2 and below" and one for "DTLS 1.3 and above". I believe this is needed (at least for the ClientHello), until version has been negotiated.

**Reply by mattcaswell** [reply]
Why is DTLS different from TLS in this regards?

**Reply by fwh-dc** [reply]
Maybe I'm wrong but the RFC states:

> In DTLS 1.3, the message transcript is computed over the
> original TLS 1.3-style Handshake messages without the
> message_seq, fragment_oﬀset, and fragment_length values. Note
> that this is a change from DTLS 1.2 where those values were
> included in the transcript.

So I guess that to allow a client to allow version negotiation with the server it must be able to include/exclude message_seq, fragment_oﬀset, and fragment_length values from the transcript depending on the selected version.

**Reply by mattcaswell** [reply]
So, can we not push everything into the mem bio and then filter those fields out later when we push them through the `handshake_dgst` (i.e. after version neg has happened)?

**Reply by fwh-dc** [reply]
Yes that is the other opportunity. Do you agree that only ClientHello needs this logic? Or are there any other messages sent before version negotiation has taken place?

**Reply by mattcaswell** [reply]
Version neg happens as a result of processing the ClientHello. However, there may (?) be some complications to consider with DTLSv1.2 HelloVerifyRequest or DTLSv1.3 HelloRetryRequest.
